tools/isobuild/import-scanner.js

```diff
@@ -0,0 +1,431 @@
+import assert from "assert";
+import {isString, has, keys, each, without} from "underscore";
+import {sha1} from "../fs/watch.js";
+import {matches as archMatches} from "../utils/archinfo.js";
+import {findImportedModuleIdentifiers} from "./js-analyze.js";
+import {
+  pathJoin,
+  pathRelative,
+  pathNormalize,
+  pathDirname,
+  pathBasename,
+  pathExtname,
+  statOrNull,
+  readFile,
+  convertToPosixPath,
+} from "../fs/files.js";
+
+export default class ImportScanner {
+  constructor({
+    name,
+    bundleArch,
+    sourceRoot,
+    extensions = [".js", ".json"],
+    usedPackageNames = {},
+    nodeModulesPath,
+  }) {
+    assert.ok(isString(sourceRoot));
+
+    this.name = name;
+    this.bundleArch = bundleArch;
+    this.sourceRoot = sourceRoot;
+    this.usedPackageNames = usedPackageNames;
+    this.nodeModulesPath = nodeModulesPath;
+    this.absPathToOutputIndex = {};
+    this.outputFiles = [];
+    this.extensions = extensions;
+  }
+
+  addInputFiles(...args) {
+    const Ap = Array.prototype;
+    const flatFiles = Ap.concat.apply(Ap, args);
+
+    flatFiles.forEach(file => {
+      const absPath = pathJoin(this.sourceRoot, file.sourcePath);
+
+      // Files that are not eagerly evaluated (lazy) will only be included
+      // in the bundle if they are actually imported.
+      file.lazy = this._isFileLazy(file);
+
+      // Files that are eagerly evaluated are effectively "imported" as
+      // entry points.
+      file.imported = ! file.lazy;
+
+      file.installPath = this._getInstallInfo(absPath).path;
+
+      if (has(this.absPathToOutputIndex, absPath)) {
```

Comment:
When does this happen? Don't we process every file once?

Or is this because someone may add the same file more than once when calling `addFiles`?